### PR TITLE
DB-11538 SYSCS_SET_GLOBAL_DATABASE_PROPERTY should check validity of options

### PR DESCRIPTION
**Pull request by martinrupp**
## Short Description
This adds a framework for giving feedback when the user calls syscs_set_global_database_property if the used value is valid (e.g. using a non-integer for an integer value). Also adding info text for options.

Note that not all options will be converted with this PR to have a validity checker.

## Long Description
currently when setting an option with `syscs_util.syscs_set_global_database_property`, we set the option to the given string value without any check. e.g. the option `splice.function.currentTimestampPrecision` requires an integer like `'6'`. However we can also set it to a non-integer like `'hello'`, without an error. Later, when the database reads that value, it won't be able to read it as an integer, so it will fall back to the default value of the option. This is also true for complicated settings like `splice.function.timestampFormat`: The format `hh:mm` is OK, however `h:mm` is not supported, and would also lead to using the default value.

## How to test
`call syscs_util.syscs_set_global_database_property('splice.function.timestampFormat', 'h:mm');`
```
ERROR XJ001: Java exception: 'java.sql.SQLException: Error parsing 'h:mm' for option splice.function.timestampFormat: java.lang.IllegalArgumentException: not supported format "h:mm": 'h' can't be repeated 1 times'.
```
`call syscs_util.syscs_set_global_database_property('splice.function.timestampFormat', 'hh:mm');`
```
name                |value                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
PROPERTY_NAME       |splice.function.timestampFormat                                                                                                                       
NEW VALUE           |hh:mm                                                                                                                                                 
PREVIOUS VALUE      |HH:mm                                                      
```

`call syscs_util.syscs_set_global_database_property('splice.function.currentTimestampPrecision', 'hello');`
```
ERROR XJ001: Java exception: 'java.sql.SQLException: Error parsing 'hello' for option splice.function.currentTimestampPrecision: java.lang.NumberFormatException: For input string: "hello"'.
```

`call syscs_util.syscs_set_global_database_property('splice.function.floatingPointNotation', 'true');`
```
ERROR XJ001: Java exception: 'java.sql.SQLException: Error parsing 'true' for option splice.function.floatingPointNotation: java.lang.RuntimeException: Supported values are [plain, normalized, default].'.
```

### CODE REVIEW

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2797/

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2844/

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2845/

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+import com.splicemachine.db.shared.common.sql.Utils;
+
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+public class GlobalDBProperties {
+    public interface Validator {
```
Comment (martinrupp) [inline db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:25]: add some documentation how to use the validator, write own validator etc.

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        void accept(String t) throws Throwable;
+    }
+
+    public static class PropertyType {
+        String name, information;
+        Validator validator;
+        public PropertyType(String name, String information, Validator validator) {
```
Comment (martinrupp) [inline db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:32]: docu

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+        public String getInformation() {
+            return information;
+        }
+
+        @Override
+        public String toString() { return name; }
+    }
+
+    public static void parseBoolean(String s) {
```
Comment (martinrupp) [inline db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:60]: validateBoolean

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+        @Override
+        public void accept(String t) throws Throwable {
+            if( Arrays.stream(options).filter( t::equalsIgnoreCase ).count() != 0)
+                return;
+            throw new RuntimeException("Supported values are " + Arrays.toString(options) + ".");
+        }
+    }
+
+    public static Stream<PropertyType> getAll() {
```
Comment (martinrupp) [inline db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:80]: docu

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                    if (!field.getType().equals(GlobalDBProperties.PropertyType.class)) return null;
+                    try {
+                        return ((PropertyType) field.get(null));
+                    } catch (IllegalAccessException e) {
+                        return null;
+                    }
+                } ).filter( f -> f != null );
+    }
+
+
```
Comment (martinrupp) [inline db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:92]: add small "how to add more properties"

**Review by martinrupp — COMMENTED** [review]
we should try to merge part of the code in e.g. `GenericStatement.setDisableParallelTaskJoinCosting`

**Comment by martinrupp** [comment]
add regex example e.g.
```
    static class RegexValidator implements Validator
    {
        Pattern pattern;

        RegexValidator(String regex) {
            pattern = Pattern.compile(regex);
        }

        @Override
        public void accept(String t) throws Throwable {
            if( !pattern.matcher(t).find())
                throw new RuntimeException("Did not match pattern " + pattern.pattern());
        }
    }

    public static PropertyType MY_INTEGER_PROPERTY =
            new PropertyType("integer.property", "an integer", new RegexValidator("\\d*");
```

**Comment by hatyo** [comment]
_I haven't done an code review yet, but,_ I want to bring your attention @martinrupp to this method `com.splicemachine.db.iapi.sql.conn.SessionProperties#validatePropertyAndValue` where we do similar validations. I think we can simplify it a bit with your new validators.

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java (hunk 1 above)
Comment (martinrupp) [reply db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:25]: done

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java (hunk 2 above)
Comment (martinrupp) [reply db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:32]: done

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java (hunk 3 above)
Comment (martinrupp) [reply db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:60]: done

**Comment by martinrupp** [comment]
> _I haven't done an code review yet, but,_ I want to bring your attention @martinrupp to this method `com.splicemachine.db.iapi.sql.conn.SessionProperties#validatePropertyAndValue` where we do similar validations. I think we can simplify it a bit with your new validators.

> we should try to merge part of the code in e.g. GenericStatement.setDisableParallelTaskJoinCosting

-> DB-12038

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+package com.splicemachine.db.iapi.reference;
+
+import com.splicemachine.db.shared.common.sql.Utils;
+
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+/**
+ * collecting class for properties that are setable with SYSCS_SET_GLOBAL_DATABASE_PROPERTY.
+ * Note: Some of the properties are still in com.splicemachine.db.iapi.reference.Property
```
[the remark this answers is not in the document]
Comment (martinrupp) [reply db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:24]: they are for all properties settable with SYSCS_SET_GLOBAL_DATABASE_PROPERTY. as written in the comment, i haven't migrated all Properties from Property to GlobalDBProperties, but that is the goal in the end.
So in short: No property should be in com.splicemachine.db.iapi.reference.Property.

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/Property.java [hunk 7]
```diff
@@ -1222,7 +1209,7 @@ Default value for wait timeouts (60 seconds)
      * How many partitions to fetch in parallel from Spark. It can increase memory consumption on RegionServer if fetching
      * too much data
      */
-    String OLAP_PARALLEL_PARTITIONS = "olapParallelPartitions";
+    String OLAP_PARALLEL_PARTITIONS = "olapParallelPartitions"; // todo: what is the difference to splice.olapParallelPartitions?
```
[the remark this answers is not in the document]
Comment (martinrupp) [reply db-shared/src/main/java/com/splicemachine/db/iapi/reference/Property.java:1212]: not clarified.

File: splice_machine/src/main/java/com/splicemachine/derby/procedures/SpliceAdmin.java [hunk 8]
```diff
@@ -1161,19 +1168,26 @@ public static void SYSCS_GET_GLOBAL_DATABASE_PROPERTIES(String filter,
...
+        propertiesList.sort(Comparator.comparing(GlobalDBProperties.PropertyType::getName));
+        for(GlobalDBProperties.PropertyType property : propertiesList) {
+            String value = PropertyUtil.getCachedDatabaseProperty(lcc, property.getName());
             if( showNonNullOnly && value == null ) continue;
             resultHelper.newRow();
-            colProperty.set(property);
+            colProperty.set(property.getName());
             colValue.set(value);
-            colInfo.set(""); // reserved to add information about properties later
+            colInfo.set(property.getInformation()); // reserved to add information about properties later
```
[the remark this answers is not in the document]
Comment (martinrupp) [reply splice_machine/src/main/java/com/splicemachine/derby/procedures/SpliceAdmin.java:1190]: done

File: splice_machine/src/main/java/com/splicemachine/derby/vti/SpliceFileVTI.java [hunk 9]
```diff
@@ -137,7 +137,7 @@ boolean getPreserveLineEndings(SpliceOperation op) throws StandardException {
             return defaultValue;
         String preserveLineEndingsString =
                 PropertyUtil.getCachedDatabaseProperty(op.getActivation().getLanguageConnectionContext(),
```
[the remark this answers is not in the document]
Comment (martinrupp) [reply splice_machine/src/main/java/com/splicemachine/derby/vti/SpliceFileVTI.java:139]: done

File: db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (martinrupp) [reply db-shared/src/main/java/com/splicemachine/db/iapi/reference/GlobalDBProperties.java:24]: i added some comments

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericStatement.java [hunk 10]
```diff
@@ -776,7 +774,7 @@ private void setCurrentTimestampPrecision(LanguageConnectionContext lcc, Compile
 
     private void setTimestampFormat(LanguageConnectionContext lcc, CompilerContext cc) throws StandardException {
         String timestampFormatString =
-                PropertyUtil.getCachedDatabaseProperty(lcc, Property.SPLICE_TIMESTAMP_FORMAT);
+                PropertyUtil.getCachedDatabaseProperty(lcc, GlobalDBProperties.SPLICE_TIMESTAMP_FORMAT.getName() );
```
[the remark this answers is not in the document]
Comment (martinrupp) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericStatement.java:777]: yes, done

**Review by ipraznik-splice — APPROVED** [review]

**Review by arnaud-splice — APPROVED** [review]
Excellent change! I think we can go one step forward in the future and keep ALL properties in the same place, tagging them with global / session / etc. or a combination of those.

**Review by carolp-503 — APPROVED** [review]
+1

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Pull request merged**